client/src/main/java/org/apache/rocketmq/client/impl/MQClientAPIImpl.java

```diff
@@ -621,7 +621,7 @@ private void onExceptionImpl(final String brokerName,
                 retryBrokerName = mqChosen.getBrokerName();
             }
             String addr = instance.findBrokerAddressInPublish(retryBrokerName);
-            log.warn(String.format("async send msg by retry {} times. topic={}, brokerAddr={}, brokerName={}", tmp, msg.getTopic(), addr,
+            log.warn(String.format("async send msg by retry %s times. topic=%s, brokerAddr=%s, brokerName=%s", tmp, msg.getTopic(), addr,
```

Comment:
Would it be better to format variable `tmp` with `%d`?

Reply:
+1

Reply:
IMHO, It would better to remove the `String.format()`, and format the exception message in another `{}`